pydantic/errors.py

```diff
@@ -0,0 +1,92 @@
+__all__ = (
+    'PydanticErrorMixin',
+    'PydanticTypeError',
+    'PydanticValueError',
+
+    'ConfigError',
+
+    'MissingError',
+    'ExtraError',
+
+    'DecimalError',
+    'DecimalIsNotFiniteError',
+    'DecimalMaxDigitsError',
+    'DecimalMaxPlacesError',
+    'DecimalWholeDigitsError',
+
+    'UUIDError',
+    'UUIDVersionError',
+)
+
+
+class PydanticErrorMixin:
+    msg_template: str
+
+    def __init__(self, **ctx):
+        self.ctx = ctx or None
+        super().__init__()
+
+    def __str__(self) -> str:
+        return self.msg_template.format(**self.ctx or {})
+
+
+class PydanticTypeError(PydanticErrorMixin,
+                        TypeError):
+    pass
+
+
+class PydanticValueError(PydanticErrorMixin,
+                         ValueError):
+    pass
+
+
+class ConfigError(RuntimeError):
+    pass
+
+
+class MissingError(PydanticValueError):
+    msg_template = 'field required'
+
+
+class ExtraError(PydanticValueError):
+    msg_template = 'extra fields not permitted'
+
+
+class DecimalError(PydanticTypeError):
+    msg_template = 'value is not a valid decimal'
```

Comment:
from gitter you said

> In your example DecimalError inherited from TypeError, so if I inherit all decimal errors from DecimalError they will be type errors not value errors as now. Is it okay?

You're right, that's not ok, so the other decimal related exceptions below can't inherit from this `DecimalError`.

However we want the end `type` on the exception to be `value_error.decimal.is_not_finite` rather than `value_error.decimal_is_not_finite`. etc.

The way to do this is to add `display_name` (or similar) to exceptions:

```py
class DecimalIsNotFiniteError(PydanticValueError):
    display_name = 'decimal.not_finite_error'
```

Then in `exceptions._get_exc_bases` do

```py
name = getattr(b, 'display_name', None) or b.__name__
yield name.replace('Error', '')
```

does that make sense?

Reply:
If we want to introduce `display_name` then I think we can simplify `get_exc_type` func logic to something like this:
```
def get_exc_type(exc: Exception) -> str:
    exc = type(exc)

    if issubclass(exc, TypeError):
        type_ = 'type_error'
    elif issubclass(exc, ValueError):
        type_ = 'value_error'
    else:
        RuntimeError(f'Unsupported exception: {exc}')

    display_name = getattr(exc, 'display_name', None)
    if display_name is not None:
        type_ = f'{type_}.{display_name}'

    return type_
```

And we can remove `to_snake_case` and all logic with MRO. Also I think it's more easy to understand and support. What do you think?

Reply:
I agree except that `f'{type_}.{display_name}'` I think will be wrong because to get `decimal.not_finite_error` and `decimal.max_digits_exceeded` you would need duplicate exception names

But maybe best for you to implement it and see how you go.